### DISCUSSION TITLE
Fresh centroid training gives precision/recall = 0, but older centroid models still work on same videos

### DISCUSSION THREAD

**Discussion by PrestonSheng** [talmolab/sleap#2812] [Help!]
Hi SLEAP Team,

I've run into an issue with my top-down SLEAP project, and I wanted to get advice on what my next steps should be.

Environment
SLEAP version: 1.6.1
sleap-nn: 0.1.0
Python: 3.13.9
PyTorch: 2.9.0+cu129
CUDA: 12.9
GPU: NVIDIA GeForce RTX 3060
OS: Windows

I have a multi-video SLEAP project with 1-2 animals per frame. I've been iteratively improving a top-down model by adding videos/labeled frames for several months now. For each training run, I trained from scratch using the selected model configuration; I did **not** resume training/fine-tune from the previous model checkpoint. The approximate timeline is

4/30-- I had 2662 labeled frames, and I trained a centroid + centered instance model. This worked.

6/30-- I had 3131 labeled frames, and I trained a centroid + centered instance model. This worked.

However, I realized here that I had labeled frames with only 1 animal labeled, despite 2 animals being in frame. I tried to clean up my project a little bit, and removed some userlabled frames, leaving me with 3019 labled frames.

7/1-- I trained a new model with this cleaned up project. It worked.

7/2-- I added more labels, reaching ~3100 labeled frames. I trained a centroid + centered instance model. **This model failed to predict any animals**

After that, I went back and tested my older models, and **only my 6/30 centroid model and before reliably predicted animals at all.**  The 7/1 and 7/2 centroid models failed to detect any animals. The centered instance model I choose does not seem to matter, which makes me think that the issue is specifically with the centroid model and centroid training.

So far, I have checked the following:

I do not think that the centroid model config changed between runs.

I did not intentionally change centroid model configs such as input scaling, max stride, etc.

The model folders look generally similar between runs.

The same videos/project can be predicted using the older 6/30 model.

The major change between the working 6/30 model and the later 7/1 model was that I removed incomplete user-labeled frames where only one of two visible animals had been labeled.

I’m currently retraining from scratch using a known-good older centroid config, but I’m not sure whether that will resolve it. I have my doubts though, since so far it looks like it's not making any progress:

Epoch 6: 100%|█████████████████████████████████████| 200/200 [04:09<00:00,  0.80it/s, loss=1.18e-04, val/loss=1.18e-04]2026-07-03 12:27:06 | Epoch 6 centroid evaluation: precision=0.0000, recall=0.0000, dist_avg=nanpx0:34<00:00,  2.30it/s]

Each model was trained from scratch. I was reusing configurations from prior models, but not resuming/fine-tuning from previous checkpoints.

Since a known older centroid model still works on the same videos, I suspect the problem is not video access or the centered-instance model, but something specific to the newer centroid training runs or the cleaned/modified training labels that is preventing training newer models.

Let me know what you guys think! 

Preston

**Edit**:

I also tried retraining a centroid model from scratch on the project version from before I removed the incomplete one-animal frames. That run also shows centroid evaluation stuck at precision/recall 0, e.g.:

Epoch 5: 100%|█████████████████████████████████████| 200/200 [03:47<00:00,  0.88it/s, loss=8.81e-05, val/loss=1.14e-04]2026-07-03 13:50:35 | Epoch 5 centroid evaluation: precision=0.0000, recall=0.0000, dist_avg=nanpx0:25<00:00,  3.34it/s]

So this may not be caused specifically by the frame-removal cleanup. It seems more like fresh centroid training is now failing on project versions where an older already-trained 6/30 centroid model still works for inference.

**Comment by PrestonSheng** [comment]
Wanted to follow this up with a shorter summary in case it helps find the issue:

The older trained centroid models from this project still work for inference on the same videos

The newly trained centroid models now fail completely, with centroid evaluation stuck at precision/recall=0 for multiple epochs.

This now also happens when training from older versions of the project. I don't think I am changing any config settings for my centroid model

The centered instance model does not seem to matter, the issue seems to be with the centroid model.

**Comment by gitttt-1234** [comment]
Hi @PrestonSheng !

My first guess is that this is a **receptive field** issue. The **Centroid model** needs a receptive field that's roughly the size of the animal in order to localize it. If it's too small, the model can't "see" enough of the animal to detect the centroid. To check this:
- Open the training dialog in the GUI and look at the receptive field preview for the centroid model (the blue box). Make sure it covers roughly the whole animal.
- If it doesn't, increase **max stride** and/or decrease **input scaling** to increase the receptive field.
- Or, try our new [config helper](https://nn.sleap.ai/latest/configuration/config-picker/). Upload your `.slp` and it'll suggest a good set of parameters tailored to your data.

One more question, **which keypoint did you set as the anchor** for the centroid model? Does it have consistently good visibility across your labeled frames? I would also suggest to upgrade to latest versions (we have a lot of new features added)

If neither of those turns out to be it, a couple of other things worth trying:
- Increase the number of filters in the model
- Share your centroid training config YAML (and if possible your slp file and video via **https://slp.sh**) so we can take a closer look.

Let us know if that works!

Thanks,

Divya

**Reply by PrestonSheng** [reply]
Hi Divya,

Thanks for the reply! I believe that my receptive field is appropriately sized for my animals. Here is a screenshot of my receptive field preview:

<img width="211" height="205" alt="image" src="https://github.com/user-attachments/assets/a0153136-a21b-4ccf-a094-2cf50c8be6f1" />

> which keypoint did you set as the anchor 

I set the "centroid" as the anchor for my centroid model, which is the point at the center of the animal. It does tend to be visible across all of my labeled frames.

One additional confusing update here: after I posted this issue, I retried training, and my centroid model randomly worked, despite me seemingly not changing anything about the model configuration. However, after labeling another round of frames and training again, the centroid model failed again to find any instances on any frames.

As far as I can tell, the working and failing centroid models used the same configuration and were both trained from scratch, so I'm confused why fresh centroid training sometimes works and sometimes completely fails. Does that still sound consistent with a receptive field or anchor issue, or would you recommend checking something else?

I'll also try updating SLEAP to the latest version. I'll upload my SLEAP project, as well as the working/not working centroid config YAMLs/model folders. 

Thank you so much!
Preston

Attached here are my centroid YAMLs.

Not working 
[initial_config.yaml](https://github.com/user-attachments/files/30194692/initial_config.yaml)
[training_config.yaml](https://github.com/user-attachments/files/30194693/training_config.yaml)

Working
[initial_config.yaml](https://github.com/user-attachments/files/30194705/initial_config.yaml)
[training_config.yaml](https://github.com/user-attachments/files/30194706/training_config.yaml)

**Reply by gitttt-1234** [reply]
Hi @PrestonSheng !

Thanks for sharing the configs and the receptive field screenshot!

The receptive field does look pretty solid, I noticed you're on an old sleap-nn version (the config says 0.0.5), however from your first post, the logs say sleap-nn 0.1.0 with SLEAP v1.6.1. Could you try [updating to the latest sleap](https://docs.sleap.ai/latest/installation/#upgrade-everything-to-latest) (this would also update all the dependencies)?

To dig further, it'd help to get:

- The videos, so we can reproduce training on our side. Rather than uploading everything (since you have more than 100 videos), you can export an embedded package: in the GUI, go to Predict → Export Labels Package… → "Labeled frames". This saves a .pkg.slp with the images embedded and you can upload that to slp.sh.
- The training logs from both a working run and a failing run if you have it. 
- How many labels were added between the working and failing runs. I wanted to confirm whether it's a bunch of new frames or a larger batch, since that could help narrow down whether a specific new instance/label is triggering it. (from run names in the configs, it looks like only one new label is added?)

Thanks,

Divya
